common/serde-reflection/src/trace.rs

```diff
@@ -2,473 +2,233 @@
-    type Error = Error;
-
-    fn serialize_element<T>(&mut self, value: &T) -> Result<()>
-    where
-        T: ?Sized + Serialize,
-    {
-        let format = value.serialize(&mut *self.tracer)?;
-        self.value.merge(format)
-    }
-
-    fn end(self) -> Result<Format> {
-        Ok(Format::Seq(Box::new(self.value)))
-    }
-}
-
-pub struct TupleTracer<'a> {
-    tracer: &'a mut Tracer,
-    values: Vec<Format>,
-}
-
-impl<'a> ser::SerializeTuple for TupleTracer<'a> {
-    type Ok = Format;
-    type Error = Error;
-
-    fn serialize_element<T>(&mut self, value: &T) -> Result<()>
-    where
-        T: ?Sized + Serialize,
-    {
-        let format = value.serialize(&mut *self.tracer)?;
-        self.values.push(format);
-        Ok(())
-    }
-
-    fn end(self) -> Result<Format> {
-        Ok(Format::Tuple(self.values))
+        loop {
+            let format = self.trace_type_once::<T>()?;
+            if let Format::TypeName(name) = &format {
+                if self.incomplete_enums.contains(name) {
+                    self.incomplete_enums.remove(name);
+                    continue;
+                }
+            }
+            return Ok(format);
+        }
     }
-}
-
-pub struct TupleStructTracer<'a> {
-    tracer: &'a mut Tracer,
-    name: &'static str,
-    fields: Vec<Format>,
-}
-
-impl<'a> ser::SerializeTupleStruct for TupleStructTracer<'a> {
-    type Ok = Format;
-    type Error = Error;
 
-    fn serialize_field<T>(&mut self, value: &T) -> Result<()>
+    /// Same as `trace` for seeded deserialization.
+    pub fn trace_type_with_seed<'de, T>(&'de mut self, seed: T) -> Result<Format>
     where
-        T: ?Sized + Serialize,
+        T: DeserializeSeed<'de> + Clone,
     {
-        let format = value.serialize(&mut *self.tracer)?;
-        self.fields.push(format);
-        Ok(())
-    }
-
-    fn end(self) -> Result<Format> {
-        let format = Format::TupleStruct(self.fields);
-        self.tracer.record(self.name, format)
+        loop {
+            let format = self.trace_type_once_with_seed(seed.clone())?;
+            if let Format::TypeName(name) = &format {
+                if self.incomplete_enums.contains(name) {
+                    self.incomplete_enums.remove(name);
+                    continue;
+                }
+            }
+            return Ok(format);
+        }
     }
-}
-
-pub struct TupleVariantTracer<'a> {
-    tracer: &'a mut Tracer,
-    name: &'static str,
-    variant_index: u32,
-    variant_name: &'static str,
-    fields: Vec<Format>,
-}
-
-impl<'a> ser::SerializeTupleVariant for TupleVariantTracer<'a> {
-    type Ok = Format;
-    type Error = Error;
 
-    fn serialize_field<T>(&mut self, value: &T) -> Result<()>
+    /// Compute a default value for `T`.
+    /// Return true if `T` is an enum and more values are needed to cover
+    /// all variant cases. Otherwise similar to `trace_type_once`.
+    pub fn sample_type_once<'de, T>(&mut self) -> Result<(T, bool)>
```

Comment:
We explore variants in sequential order. This matters for termination that we start where the base case(s) should be.

Reply:
We never fail to produce a value (as long as we terminate and) unless a user-defined deserialize trait raises an exception. In the latter case, it would return an `Error::Custom` here.